gt/main.c

```diff
@@ -736,6 +820,73 @@ alloc_and_fill_notify_pkt(unsigned int socket, struct ggu_policy *policy,
 	return notify_pkt;
 }
 
+/*
+ * Use the @dr to notify the GK
+ * about the punishment policies on declined flows
+ * with fragmented packets.
+ */
+static void 
+process_death_row(int socket_id, uint8_t port, uint16_t tx_queue,
+	struct rte_ip_frag_death_row *death_row,
+	struct gt_instance *instance, struct gt_config *gt_conf)
+{
+	uint32_t i;
+
+	for (i = 0; i < death_row->cnt; i++) {
+		struct gt_packet_headers pkt_info;
+		struct ggu_policy policy;
+		struct rte_mbuf *notify_pkt;
+		int ret = gt_parse_incoming_pkt(death_row->row[i], &pkt_info);
+		if (ret < 0) {
+			rte_pktmbuf_free(death_row->row[i]);
+			continue;
+		}
+
+		ret = lookup_frag_punish_policy_decision(
+			&pkt_info, &policy, instance);
+		if (ret < 0) {
+			rte_pktmbuf_free(death_row->row[i]);
+			continue;
+		}
+
+		/*
+		 * TODO Reply in a batch.
+		 * Reply the policy decision to GK-GT unit.
+		 */
+		notify_pkt = alloc_and_fill_notify_pkt(
+			socket_id, &policy, &pkt_info, gt_conf);
+		if (notify_pkt != NULL && rte_eth_tx_burst(
+				port, tx_queue, &notify_pkt, 1) != 1)
+			rte_pktmbuf_free(notify_pkt);
```

Comment:
We must log that a notification was not sent out. Report the policy decision in the message.

Reply:
Should we report all the information of the policy decision? e.g., the IP flow information.

Reply:
Yes.

Reply:
The idea is that we want to help network administrators to deal with possibly bad situations.